src/xvector/nnet-xvector-compute.cc

```diff
@@ -0,0 +1,116 @@
+// xvector/nnet-xvector-compute.cc
+
+// Copyright      2015    Johns Hopkins University (author: Daniel Povey)
+//                2016    David Snyder
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "xvector/nnet-xvector-compute.h"
+#include "nnet3/nnet-utils.h"
+
+namespace kaldi {
+namespace nnet3 {
+
+NnetXvectorComputer::NnetXvectorComputer(
+    const NnetSimpleComputationOptions &config,
+    Nnet *nnet):
+    nnet_(nnet),
+    config_(config),
+    compiler_(*nnet, config.optimize_config) {
+}
+
+void NnetXvectorComputer::ComputeXvector(const MatrixBase<BaseFloat> &feats,
+                    Vector<BaseFloat> *xvector) {
+
+  ComputationRequest request;
+  GetComputationRequest(feats, &request);
+  const NnetComputation *computation = compiler_.Compile(request);
+  NnetComputer computer(config_.compute_config, *computation,
+                        *nnet_,
+                        nnet_);
+  std::string input_name = "input";
+  CuMatrix<BaseFloat> cu_feats(feats);
+  computer.AcceptInput(input_name, &cu_feats);
+  computer.Forward();
+  const CuMatrixBase<BaseFloat> &output = computer.GetOutput("output");
+  KALDI_ASSERT(output.NumRows() == 1 && output.NumCols() == xvector->Dim());
+  CuSubVector<BaseFloat> xvector_tmp(output, 0);
+  xvector->CopyFromVec(xvector_tmp);
+}
+
+void NnetXvectorComputer::GetComputationRequest(
+    const MatrixBase<BaseFloat> &feats,
+    ComputationRequest *request) {
+  std::string input_name = "input",
+              output_name = "output";
+  NnetIo nnet_io = NnetIo(input_name, 0, feats);
+  request->inputs.clear();
+  request->outputs.clear();
+  request->inputs.resize(1);
+  request->outputs.resize(1);
+  request->need_model_derivative = false;
+  request->store_component_stats = false;
+
+  int32 input_node_index = nnet_->GetNodeIndex(input_name);
+
+  if (input_node_index == -1 && !nnet_->IsInputNode(input_node_index))
+    KALDI_ERR << "No input node called '" << input_name
+              << "' in the network.";
+
+  request->inputs[0].name = input_name;
+  request->inputs[0].indexes = nnet_io.indexes;
+  request->inputs[0].has_deriv = false;
+
+  // We only need the output on frame t=0.
+  int32 io_index_size = request->inputs[0].indexes.size(),
+         n_indx_size = 1e6, t_ind;
```

Comment:
BTW, in nnet3-xvector-compute.cc, here
            xvector_mat.Row(i).CopyFromVec(xvector);
        } else
          xvector_mat.Row(chunk_indx).CopyFromVec(xvector);
      }

there should be braces on the else since there were on the original if.


On Tue, Mar 1, 2016 at 4:15 PM, david-ryan-snyder <notifications@github.com>
wrote:

> In src/xvector/nnet-xvector-compute.cc
> <https://github.com/danpovey/kaldi/pull/21#discussion_r54636729>:
>
> > +  request->need_model_derivative = false;
> > +  request->store_component_stats = false;
> > +
> > +  int32 input_node_index = nnet_->GetNodeIndex(input_name);
> > +
> > +  if (input_node_index == -1 && !nnet_->IsInputNode(input_node_index))
> > +    KALDI_ERR << "No input node called '" << input_name
> > +              << "' in the network.";
> > +
> > +  request->inputs[0].name = input_name;
> > +  request->inputs[0].indexes = nnet_io.indexes;
> > +  request->inputs[0].has_deriv = false;
> > +
> > +  // We only need the output on frame t=0.
> > +  int32 io_index_size = request->inputs[0].indexes.size(),
> > +         n_indx_size = 1e6, t_ind;
>
> I just sent a pull request to simplify that.
> … <#833163982_>
> On Tue, Mar 1, 2016 at 4:13 PM, Daniel Povey <notifications@github.com>
> wrote: In src/xvector/nnet-xvector-compute.cc <#21 (comment)
> <https://github.com/danpovey/kaldi/pull/21#discussion_r54636395>>: > +
> request->need_model_derivative = false; > + request->store_component_stats
> = false; > + > + int32 input_node_index = nnet_->GetNodeIndex(input_name);
> > + > + if (input_node_index == -1 &&
> !nnet_->IsInputNode(input_node_index)) > + KALDI_ERR << "No input node
> called '" << input_name > + << "' in the network."; > + > +
> request->inputs[0].name = input_name; > + request->inputs[0].indexes =
> nnet_io.indexes; > + request->inputs[0].has_deriv = false; > + > + // We
> only need the output on frame t=0. > + int32 io_index_size =
> request->inputs[0].indexes.size(), > + n_indx_size = 1e6, t_ind; the code
> starting around here is very strange and complex. I don't understand why
> all this is needed. — Reply to this email directly or view it on GitHub <
> https://github.com/danpovey/kaldi/pull/21/files#r54636395>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/danpovey/kaldi/pull/21/files#r54636729>.
>